src/transformStateWithClones.js

```diff
@@ -7,7 +7,28 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let currentState = { ...state };
+  const stateHistory = [];
+
+  for (const action of actions) {
+    if (action.type === 'clear') {
+      currentState = {};
+    }
+
+    if (action.type === 'removeProperties') {
+      for (const key of action.keysToRemove) {
+        delete currentState[key];
+      }
+    }
+
+    if (action.type === 'addProperties') {
+      Object.assign(currentState, action.extraData);
+    }
+    stateHistory.push({ ...currentState });
```

Comment:
Pushing a copy of the object to the array should be done at the end of the loop cycle but outside of the switch block to avoid repeating this action in every case.